### PR TITLE
mrc-4561 implement authentication 

### PR DESCRIPTION
**Pull request by LekanAnni**
This PR serves as a proof of concept for implementing basic packet authentication. It lays the foundation for future enhancements, including token validation, fine-grained access control, and additional business logic as needed. I'm seeking a preliminary review before proceeding with test implementation.

TODO: 
1. Add browser test
2. Add logout endpoint in backend
3. Call logout endpoint in front end to destroy backend securityContext
4. Return custom message for unauthenticated handler in backend

To test:

1. Create a GitHub App
2. Use the below config
3. update api config.property
5. set GitHub secret and client id
6. run app

See below for a screenplay and GitHub config 

Screenshot 2023-09-14 at 18 17 14

https://github.com/mrc-ide/packit/assets/4240748/fc73aa74-5200-41d5-9b2e-41bce27f85f9

### CODE REVIEW

File: api/app/src/main/kotlin/packit/security/profile/TokenToPrincipalConverter.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            "",
+            extractAuthorities(jwt),
+            jwt.getClaim("name").toString(),
+            mutableMapOf()
+        )
+    }
+
+    override fun extractAuthorities(jwt: DecodedJWT): MutableCollection<out GrantedAuthority>
+    {
+        val claims = jwt.getClaim("a")
```
Comment (EmmaLRussell) [inline api/app/src/main/kotlin/packit/security/profile/TokenToPrincipalConverter.kt:30]: How does this get set?

File: api/app/src/main/kotlin/packit/security/oauth2/OAuth2SuccessHandler.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    override fun handle(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        authentication: Authentication,
+    )
+    {
+        var targetUrl =
+            config.authRedirectUri.ifEmpty { determineTargetUrl(request, response, authentication) }
+
+        val token = jwtIssuer.issue(authentication)
```
Comment (EmmaLRussell) [inline api/app/src/main/kotlin/packit/security/oauth2/OAuth2SuccessHandler.kt:37]: So we still want to use a JWT token for OAuth2 users (encodes roles etc) but that token  will be signed with the secret we configured for basic auth? 
https://github.com/mrc-ide/packit/pull/29/files#diff-3cfc73e7540bb99e4da3587ab2e12ea5eb950c8fc6b92dae45dc99305717e8ccR40

**Review by EmmaLRussell — COMMENTED** [review]
Got this working for github access. :+1: 

We aren't going to use Basic auth, only Github or montagu, so I don't think it's that useful to include the Basic option. I guess it's good to be able to show that the app can support multiple methods, but that will be at the level of app configuration anyway, not something the user chooses on login, 

> I'm seeking a preliminary review before proceeding with test implementation.

This is a pretty complex PR on tech that I don't think any of us are familiar with (spring boot auth), so to get the most out of your reviewers, it would probably be helpful if you could expand a little on the key parts of the code or any parts which you feel might be contentious.

I've really only looked at the backend code so far, but I think that's where the complexity is.

File: api/app/src/main/kotlin/packit/service/UserService.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import packit.security.Role
+import java.util.*
+
+interface UserService
+{
+    fun findByEmail(email: String): Optional<User>
+}
+
+@Service
+class UserServiceImpl : UserService
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/service/UserService.kt:14]: I will leave this for now.

File: api/app/src/main/kotlin/packit/security/oauth2/Oauth2UserService.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        //TODO check if user exists, if not, save user email to database
+
+        val user = User(
+            1L,
+            githubInfo.email(),
+            "",
+            Role.USER,
+            githubInfo.name(),
+            oAuth2User.attributes
+        )
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/oauth2/Oauth2UserService.kt:40]: Yes, we would perform different fine-grain based permission.

File: api/app/src/main/kotlin/packit/security/oauth2/GithubOAuth2UserInfo.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    override fun email(): String
+    {
+        return attributes["email"].toString()
+    }
+
+    override fun imageUrl(): String
+    {
+        return attributes["avatar_url"].toString()
+    }
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/oauth2/GithubOAuth2UserInfo.kt]: removed

File: api/app/src/main/kotlin/packit/controllers/PacketController.kt [hunk 6]
```diff
@@ -15,6 +17,7 @@ class PacketController(private val packetService: PacketService)
 {
     @GetMapping
     fun pageableIndex(
+        @AuthenticationPrincipal userPrincipal: UserPrincipal,
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/controllers/PacketController.kt]: This was for debugging, I wanted to know the current authenticated user. Will be removed.

File: api/app/src/main/kotlin/packit/security/profile/TokenToPrincipalConverter.kt (hunk 1 above)
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/profile/TokenToPrincipalConverter.kt]: I have now updated this, meant to be "au" for authorities.

File: api/app/src/main/kotlin/packit/security/oauth2/OAuth2SuccessHandler.kt (hunk 2 above)
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/oauth2/OAuth2SuccessHandler.kt:37]: Yes, we are issuing the token because app is stateless.

File: api/app/src/main/kotlin/packit/model/User.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package packit.model
+
+import net.minidev.json.annotate.JsonIgnore
+import packit.security.Role
+
+data class User(
+    val id: Long,
+    val email: String,
+    @JsonIgnore
+    val password: String,
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/model/User.kt:10]: I will remove the password when we integrate Montagu to the OAuth2.

File: app/src/app/components/routes/Router.tsx [hunk 8]
```diff
@@ -7,21 +7,25 @@ import {NotFound} from "../NotFound";
...
-                <Route path="packets/:packetId/changelogs" element={<ChangeLogs/>}/>
+                <Route index element={<ProtectedRoute element={<Explorer/>}/>}/>
+                <Route path="packets" element={<ProtectedRoute element={<Explorer/>}/>}/>
+                <Route path="run" element={<ProtectedRoute element={<PacketRunner/>}/>}/>
+                <Route path="run-workflow" element={<ProtectedRoute element={<WorkflowRunner/>}/>}/>
+                <Route path="documentation" element={<ProtectedRoute element={<ProjectDocumentation/>}/>}/>
+                <Route path="packets/:packetId" element={<ProtectedRoute element={<PacketDetails/>}/>}/>
+                <Route path="packets/:packetId/metadata" element={<ProtectedRoute element={<Metadata/>}/>}/>
+                <Route path="packets/:packetId/downloads" element={<ProtectedRoute element={<Download/>}/>}/>
+                <Route path="packets/:packetId/changelogs" element={<ProtectedRoute element={<ChangeLogs/>}/>}/>
```
Comment (M-Kusumgar) [inline app/src/app/components/routes/Router.tsx:28]: Since you are just doing protected route in all of these, you could just use a route element that wraps all of these and just have the protected route logic in there with an `<Outlet/>` if they are authenticated?

**Review by M-Kusumgar — COMMENTED** [review]
I was able to follow a decent amount of the backend but I mainly looked at the frontend code. I am not experienced enough in auth to give improvements to backend but it looked alright. Just a small comment about the frontend but other than that, it looks good to me!

File: app/src/app/components/routes/Router.tsx (hunk 8 above)
Comment (LekanAnni) [reply app/src/app/components/routes/Router.tsx]: Done

File: api/app/src/test/kotlin/packit/integration/controllers/OutpackControllerTest.kt [hunk 9]
```diff
@@ -102,17 +115,19 @@ class OutpackControllerTest : IntegrationTest()
...
+            "/outpack/packet/badhash",
+            getTokenizedHttpEntity(MediaType.TEXT_PLAIN, testPacket),
+            String::class.java
+        )
 
         assertEquals(result.statusCode, HttpStatusCode.valueOf(400))
         jsonValidator.validateError(
                 result.body!!, "invalid input parameter",
-                "Hash of packet does not match"
+                "Invalid hash format 'badhash'"
```
Comment (LekanAnni) [inline api/app/src/test/kotlin/packit/integration/controllers/OutpackControllerTest.kt:130]: Changes came from outback server

File: api/scripts/smoke-test [hunk 10]
```diff
@@ -5,7 +5,7 @@ wait_for()
     start_ts=$(date +%s)
     for i in $(seq $TIMEOUT); do
 
-        result=$(curl --write-out %{http_code} --silent --output /dev/null http://localhost:8080/packets)
```
Comment (LekanAnni) [inline api/scripts/smoke-test:8]: /packets is not protected so I changed to unprotected route.

File: app/src/apiService.ts [hunk 11]
```diff
@@ -1,20 +1,53 @@
...
 import {Error} from "./types";
 import appConfig from "./config/appConfig";
+import {validateToken} from "./helpers";
+import {getCurrentUser} from "./localStorageManager";
+import {Store} from "@reduxjs/toolkit";
+
+let store: any;
+
+export const injectStore = (_store: any) => {
+    store = _store;
```
Comment (LekanAnni) [inline app/src/apiService.ts:11]: I could not test this line

File: app/src/apiService.ts [hunk 12]
```diff
@@ -1,20 +1,53 @@
...
         this.axiosInstance = axiosInstance;
+        this.injectedStore = injectStore;
     }
 
+    addAuthorizationHeader = (axiosInstance: AxiosInstance) => {
+
+        axiosInstance.interceptors.request.use(config => {
+            if (validateToken(getCurrentUser() || this.injectedStore.getState().login.user)) {
+                config.headers.authorization = `Bearer ${getCurrentUser().token
+                || this.injectedStore.getState().login.user.token}`;
```
Comment (LekanAnni) [inline app/src/apiService.ts:45]: the same

File: app/src/app/components/header/HeaderDropDown.tsx [hunk 13]
```diff
@@ -2,19 +2,36 @@ import React from "react";
...
-            <NavDropdown.Item>
-                <LogoutIcon className="me-2" sx={{fontSize: 20}}/>
-                Logout
-            </NavDropdown.Item>
-        </NavDropdown>
+        <>
+            {isAuthenticated &&
+                <NavDropdown data-testid="drop-down" className="icon-primary" title={
+                    <AccountCircleIcon sx={{fontSize: 35}}/>}>
+                    <NavDropdown.Item>l.ani@imperial.ac.uk</NavDropdown.Item>
```
Comment (EmmaLRussell) [inline app/src/app/components/header/HeaderDropDown.tsx:25]: Could replace this with the real user name or email now I guess?

**Review by EmmaLRussell — COMMENTED** [review]
LGTM, just got a few questions which we can discuss tomorrow. 

It looks to me like this approach should work for both web app users, and those accessing the endpoint directly over the API, since the issued token can be used in both cases i.e we're not using a cookie. Is that right? UPDATE: yes, confirmed at today's standup, though some modification to be done in a separate ticket.

I think there's some potential security concerns with storing tokens in local storage but the same goes for using a cookie, and i think it should be ok for our purposes.. 

It would be good to get in an integration test for github login, though that's obviously going to be a lot trickier than for the basic auth path. In orderly-web we used hoverfly to spoof the oauth2 handshake from the github side - there might be nicer ways of doing that these days... 

I'm a little bit unclear on how the redirection in the backend interacts with the frontend, but let's talk about that tomorrow.

File: api/app/src/main/kotlin/packit/security/WebSecurityConfig.kt [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+            .addFilterBefore(tokenAuthenticationFilter, UsernamePasswordAuthenticationFilter::class.java)
+            .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+            .and()
+            .formLogin().disable()
+            .securityMatcher("/**")
+            .authorizeHttpRequests { authorizeRequests ->
+                authorizeRequests
+                    .requestMatchers("/").permitAll()
+                    .requestMatchers("/auth/**", "/oauth2/**").permitAll()
+                    .anyRequest().authenticated()
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/WebSecurityConfig.kt:50]: That's right, so we can get to` /auth/login` for example.

File: api/app/src/main/kotlin/packit/security/WebSecurityConfig.kt [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        return httpSecurity.build()
+    }
+
+    @Bean
+    fun passwordEncoder(): PasswordEncoder
+    {
+        return BCryptPasswordEncoder()
+    }
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/WebSecurityConfig.kt:74]: That's correct.

File: api/app/src/main/kotlin/packit/security/oauth2/OAuth2SuccessHandler.kt [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        authentication: Authentication,
+    )
+    {
+        val token = jwtIssuer.issue(authentication)
+
+        val targetUrl = buildUri(token)
+
+        redirectStrategy.sendRedirect(request, response, targetUrl)
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/oauth2/OAuth2SuccessHandler.kt:38]: Not really, after successful login attempt, we will be redirecting user to the frontend adding access token as query param.

File: api/app/src/main/kotlin/packit/model/User.kt [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+    val role: Role,
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/model/User.kt:11]: Yes I agree, perhaps when we start implementing fine grained.

File: api/app/src/main/kotlin/packit/security/provider/TokenProvider.kt [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+    override fun issue(authentication: Authentication): String
+    {
+        val user = authentication.principal as UserPrincipal
+
+        val roles = user.authorities.map { it.authority }
+
+        val createdDate = Instant.now()
+
+        val expiredDate = createdDate.plus(Duration.of(config.authExpiryDays, ChronoUnit.DAYS))
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/provider/TokenProvider.kt:35]: We can add refresh token implementation to the solution,  users will have to re-login when token expires. Credentials do not expire because it's for username and password => basic user login method.

File: api/app/src/main/kotlin/packit/security/profile/UserPrincipalAuthenticationToken.kt [hunk 19]
```diff
@@ -0,0 +1,24 @@
+package packit.security.profile
+
+import org.springframework.security.authentication.AbstractAuthenticationToken
+
+class UserPrincipalAuthenticationToken(userPrincipal: UserPrincipal) :
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/main/kotlin/packit/security/profile/UserPrincipalAuthenticationToken.kt:5]: It is needed for extracting token from user principal which stores user details. We might be able to rid of it when we remove support for basic login method

File: api/app/src/test/kotlin/packit/integration/controllers/LoginControllerTest.kt [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+        val headers = HttpHeaders()
+
+        headers.contentType = MediaType.APPLICATION_JSON
+
+        val requestEntity = HttpEntity(loginRequest, headers)
+
+        val result: ResponseEntity<String> = restTemplate.exchange("/auth/login", HttpMethod.POST, requestEntity)
+
+        assertSuccess(result)
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply api/app/src/test/kotlin/packit/integration/controllers/LoginControllerTest.kt:28]: This assertSuccess(result) already asserts that the body is not empty which is the token. I also have an integration test in frontend asserting token is not empty here  app/src/tests/integration/login/actions.test.tsx.

File: app/src/app/components/contents/explorer/Table.tsx [hunk 21]
```diff
@@ -43,7 +43,7 @@ export default function Table({data}: PacketTableProps) {
                 <tr key={`row-${key}`}>
                     <td>
                         <span>
-                            <Link to={`/packets/${packet.id}`}>
+                            <Link to={`/packets/${packet.id}/packet`}>
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply app/src/app/components/contents/explorer/Table.tsx:46]: both endpoints function as same, This update ensures navLink is highlighted for selected tab.

File: demos/orderly/.outpack/files/sha256/9f/86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08 [hunk 22]
```diff
@@ -0,0 +1 @@
+test
```
[the remark this answers is not in the document]
Comment (LekanAnni) [reply demos/orderly/.outpack/files/sha256/9f/86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08:1]: I am not sure why this changes occur but I guess it was initiated from outback-server config  as changes were made there.

File: app/src/app/components/header/HeaderDropDown.tsx (hunk 13 above)
Comment (LekanAnni) [reply app/src/app/components/header/HeaderDropDown.tsx:25]: Oh yes, I will vote to do that in a different ticket. https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4638/Validate-and-extract-user-details-from-Token

File: app/src/apiService.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (LekanAnni) [reply app/src/apiService.ts:45]: Tried reading from localStorage alone but reactivity issue with local storage made that impossible to work properly. Can move to util here https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4638/Validate-and-extract-user-details-from-Token

**Review by EmmaLRussell — COMMENTED** [review]
Looks good, and thanks for the walkthrough. As discussed, would be great to get the no auth config option in before merging this.

File: api/app/src/main/kotlin/packit/controllers/LoginController.kt [hunk 23]
```diff
@@ -9,6 +9,7 @@ import packit.service.UserLoginService
 @RestController
 @RequestMapping("/auth")
 class LoginController(val loginService: UserLoginService)
+
```
Comment (EmmaLRussell) [inline api/app/src/main/kotlin/packit/controllers/LoginController.kt:12]: unnecessary empty line?

**Review by EmmaLRussell — COMMENTED** [review]
I got a backend error when I ran this up the first time, related to an old token:
```
java.lang.RuntimeException: packit.exceptions.PackitException: PackitException with key expired token: The Token has expired on 2023-10-05T11:07:40Z.
        at org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter.doFilterInternal(OAuth2AuthorizationRequestRedirectFilter.java:212) ~[spring-security-oauth2-client-6.0.1.jar:6.0.1]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:107) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:93) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:91) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.context.SecurityContextHolderFilter.doFilter(SecurityContextHolderFilter.java:82) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.context.SecurityContextHolderFilter.doFilter(SecurityContextHolderFilter.java:69) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:62) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.session.DisableEncodeUrlFilter.doFilterInternal(DisableEncodeUrlFilter.java:42) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:233) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:191) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:351) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267) ~[spring-web-6.0.4.jar:6.0.4]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:119) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:400) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:859) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1734) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-10.1.5.jar:10.1.5]
        at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: packit.exceptions.PackitException: PackitException with key expired token: The Token has expired on 2023-10-05T11:07:40Z.
        at packit.security.provider.TokenDecoder.decode(TokenDecoder.kt:33) ~[main/:na]
        at packit.security.TokenAuthenticationFilter$doFilterInternal$1.invoke(TokenAuthenticationFilter.kt:33) ~[main/:na]
        at packit.security.TokenAuthenticationFilter$doFilterInternal$1.invoke(TokenAuthenticationFilter.kt:33) ~[main/:na]
        at packit.security.TokenAuthenticationFilter.doFilterInternal$lambda$0(TokenAuthenticationFilter.kt:33) ~[main/:na]
        at java.base/java.util.Optional.map(Optional.java:260) ~[na:na]
        at packit.security.TokenAuthenticationFilter.doFilterInternal(TokenAuthenticationFilter.kt:33) ~[main/:na]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.0.4.jar:6.0.4]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:227) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:221) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374) ~[spring-security-web-6.0.1.jar:6.0.1]
        at org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter.doFilterInternal(OAuth2AuthorizationRequestRedirectFilter.java:181) ~[spring-security-oauth2-client-6.0.1.jar:6.0.1]
```
In the front end, I could see the main packets page, but with no content:
image

This isn't anything to do with the change to enable no auth, but seems to be a general issue with handling expired tokens, I'd be fine with sorting this in a separate ticket, or this one.

Worked fine with cleared local storage, though still seeing hardcoded uesr name (I thought you'd taken that out, did it creep back in?)
image

The no auth option works fine, user is not redirected to the login page. However you can browse directly to /login and see the login options. We should redirect to / in this case. Again, I think this can go in a separate ticket (not high priority as you wouldn't expect users to go to /login for a no auth instance).

**Comment by LekanAnni** [comment]
> Worked fine with cleared local storage, though still seeing hardcoded uesr name (I thought you'd taken that out, did it creep back in?) image
> 
> The no auth option works fine, user is not redirected to the login page. However you can browse directly to /login and see the login options. We should redirect to / in this case. Again, I think this can go in a separate ticket (not high priority as you wouldn't expect users to go to /login for a no auth instance).

Ticket created here already https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4638/Validate-and-extract-user-details-from-Token

**Comment by LekanAnni** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> In the front end, I could see the main packets page, but with no content:
> image

401 is now being handled in frontend, redirects to login page.

**Review by EmmaLRussell — APPROVED** [review]
Looks good to me. happy for this to be merged. 

I've made a ticket for redirecting from /login in unauthenitcated mode - it's unlikely to be an issue but could potentially be confusing: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4656

File: app/src/app/components/login/Redirect.tsx [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+        }
+    }, [token]);
+
+    if (!isAuthenticated) {
+        return (<div><p>Redirecting...</p></div>);
+    }
+    return (
+        <div>
+            <p>Redirecting user</p>
+            {<Navigate to="/"/>}
```
Comment (M-Kusumgar) [inline app/src/app/components/login/Redirect.tsx:27]: dont think you need those
```suggestion
            <Navigate to="/"/>
```

**Review by M-Kusumgar — APPROVED** [review]
had a second look at this PR, happy to approve it!
